### PR TITLE
feat(@leav/ui): improve explorer display  performances

### PR DESCRIPTION
**Pull request by evoiron**
## Checklist

Definition Of Review

-   [x] Own code review done (add notes for others)
-   [ ] Write message in teams channel
    ```
     <Title>

    *️⃣ Impacted projects : Core - DataStudio - Admin - @leav/ui - @leav/utils - ...

    📖 Ticket: https://aristid.atlassian.net/browse/<JIRA_TICKET_IDENTIFIER>

    🧑‍💻 PR: <link to PR/MR>

    ℹ Info: <brief explanation - context - how to test>
    ```

Definition Of Mergeable

-   [ ] 2 approves
-   [ ] 1 functional review - US has been tested
-   [ ] Every comment is handled - blocking ones have been resolved by reviewer
-   [ ] Design OK
-   [ ] Can be tested by POs
-   [ ] PR was introduced during daily meeting

### CODE REVIEW

File: libs/ui/src/components/Explorer/Explorer.tsx [hunk 1]
```diff
@@ -102,18 +106,24 @@ export const Explorer: FunctionComponent<IExplorerProps> = ({
...
-                  pageSize: view.pageSize,
-                  setNewPageSize,
-                  setNewPage,
-                  totalItems: data?.totalCount ?? 0
-              }
-          };
+    const dataViewProps: Pick<
+        ComponentProps<typeof DataView>,
+        'paginationProps' | 'attributesToDisplay' | 'itemActions'
+    > = useMemo(
```
Comment (evoiron) [inline libs/ui/src/components/Explorer/Explorer.tsx:112]: **note:** some dataView properties are considered changed although values didn't change (only some references did), so using memoization helps

File: libs/ui/src/components/Explorer/Explorer.tsx [hunk 2]
```diff
@@ -26,9 +26,13 @@ import {
 import {useSearchInput} from './useSearchInput';
 import {usePagination} from './usePagination';
 import {Loading} from '../Loading';
+import {Empty} from 'antd';
 
 const isNotEmpty = <T extends unknown[]>(union: T): union is Exclude<T, []> => union.length > 0;
 
+const emptyArray = [];
```
Comment (evoiron) [inline libs/ui/src/components/Explorer/Explorer.tsx:33]: **note:** [] or {} should not be used directly in components properties as they lead to new reference at each render. Also true if declared as const inside the component. The solution is to declare as constants outside, so the reference is always the same

File: libs/ui/src/components/Explorer/_queries/useExplorerData.ts [hunk 3]
```diff
@@ -69,8 +70,10 @@ export const useExplorerData = ({
         }
     });
 
+    const memoizedData = useMemo(() => (data !== undefined ? _mapping(data, libraryId, availableLangs) : null), [data]);
```
Comment (evoiron) [inline libs/ui/src/components/Explorer/_queries/useExplorerData.ts:73]: **note:** each call to _mapping return a new object, even if data have not changed

File: libs/ui/src/components/Explorer/useEditAction.tsx [hunk 4]
```diff
@@ -36,7 +36,7 @@ export const useEditAction = ({isEnabled}: ActionHook) => {
     const _handleEditModalClose = (item: IItemData) => {
         refreshItem({
             variables: {
-                libraryId: item.libraryId,
+                libraryId: editingItem?.libraryId ?? '',
```
Comment (P0ppoff) [inline libs/ui/src/components/Explorer/useEditAction.tsx:39]: **question:** why ?

File: libs/ui/src/components/Explorer/Explorer.tsx [hunk 5]
```diff
@@ -123,6 +133,7 @@ export const Explorer: FunctionComponent<IExplorerProps> = ({
                         <ExplorerTitle library={library} title={title} />
                     </KitTypography.Title>
                     <KitSpace size="xs">
+                        kikoo
```
Comment (P0ppoff) [inline libs/ui/src/components/Explorer/Explorer.tsx:136]: **todo:** remove

File: libs/ui/src/components/Explorer/Explorer.tsx [hunk 6]
```diff
@@ -102,18 +106,24 @@ export const Explorer: FunctionComponent<IExplorerProps> = ({
...
+                currentPage,
+                pageSize: view.pageSize,
+                setNewPageSize,
+                setNewPage,
+                totalItems: data?.totalCount ?? 0
+            },
+            itemActions: [editAction, deactivateAction, ...(itemActions ?? emptyArray)].filter(Boolean)
+        }),
+        [view.attributesIds, data?.totalCount, itemActions, editAction, deactivateAction]
+    );
```
Comment (P0ppoff) [inline libs/ui/src/components/Explorer/Explorer.tsx:126]: **todo:** à essayer : https://react.dev/reference/react/memo#memo avec la fonction `arePropsEqual`

File: libs/ui/src/components/Explorer/DataView.tsx [hunk 7]
```diff
@@ -47,128 +47,136 @@ interface IDataViewProps {
...
-    attributesToDisplay,
-    attributesProperties,
-    paginationProps,
-    itemActions
-}) => {
-    const {t} = useSharedTranslation();
-    const {theme} = useKitTheme();
+const arePropsEqual = (prevProps: IDataViewProps, nextProps: IDataViewProps) =>
+    prevProps.itemActions.length === nextProps.itemActions.length &&
+    prevProps.attributesToDisplay.length === nextProps.attributesToDisplay.length &&
```
Comment (P0ppoff) [inline libs/ui/src/components/Explorer/DataView.tsx:52]: **question:** Si tu changes l'ordre ? c'est la même longueur mais pas le même contenu ?

Comment (P0ppoff) [reply libs/ui/src/components/Explorer/DataView.tsx:52]: En fait je pense que l'on peut résumer en : il faut rendre le composant à chaque changement sur les data uniquement. Peut être que mettre un `arePropsEqual` sur `dataGroupedFilteredSorted` ça peut être bien

File: libs/ui/src/components/Explorer/Explorer.tsx [hunk 8]
```diff
@@ -1,7 +1,7 @@
 // Copyright LEAV Solutions 2017 until 2023/11/05, Copyright Aristid from 2023/11/06
 // This file is released under LGPL V3
 // License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
-import {ComponentProps, FunctionComponent, useReducer} from 'react';
+import {ComponentProps, FunctionComponent, useMemo, useReducer} from 'react';
```
Comment (P0ppoff) [inline libs/ui/src/components/Explorer/Explorer.tsx:4]: **typo:** tu n'en as plus besoin
```suggestion
import {ComponentProps, FunctionComponent, useReducer} from 'react';
```

File: libs/ui/src/components/Explorer/Explorer.tsx [hunk 9]
```diff
@@ -28,9 +28,13 @@ import {usePagination} from './usePagination';
 import {Loading} from '../Loading';
 import {ExplorerFilterBar} from './display-view-filters/ExplorerFilterBar';
 import {useSharedTranslation} from '_ui/hooks/useSharedTranslation';
+import {Empty} from 'antd';
```
Comment (P0ppoff) [inline libs/ui/src/components/Explorer/Explorer.tsx:31]: **todo:** not used too

**Review by P0ppoff — APPROVED** [review]

**Review by TdyP — APPROVED** [review]

File: libs/ui/src/components/Explorer/DataView.tsx (hunk 7 above)
Comment (evoiron) [reply libs/ui/src/components/Explorer/DataView.tsx:52]: euh bin non ça ne marche pas

**Pull request merged**